### PR TITLE
chore: add methods to conditionally remove stackset instance from app

### PR DESCRIPTION
**Pull request by bvtujo**
PR 1 of 2 replacing #4992.

This adds support methods and a whole boatload of mocks and unit tests to the stackset client and Cloudformation packages. These are necessary to conditionally remove the stackset instance, or redeploy the stackset instances & app to remove an account from dns delegation and IAM roles.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/aws/cloudformation/stackset/stackset.go
Comment (iamhopaul123) [inline internal/pkg/aws/cloudformation/stackset/stackset.go]:
```suggestion
		return "", fmt.Errorf("delete stack instance in region %v for account %v for stackset %s: %w",
```
I guess we only have one stack instance?

File: internal/pkg/deploy/cloudformation/app.go
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/app.go:386]: Why don't we move the logic for determining the value for "DeleteStackInstance" and "RemoveAccountFromApp" here in this function? So as to make sure calling `RemoveEnvFromApp` is always safe.

**Review by iamhopaul123 — COMMENTED** [review]
This is awesome!

File: internal/pkg/deploy/cloudformation/app.go
Comment (bvtujo) [reply internal/pkg/deploy/cloudformation/app.go:386]: We'll have to duplicate this logic somewhere, though, because we have to check in env_delete.go whether or not to empty the s3 buckets and ecr repos. I don't love that, but I see the advantage of having logic in the RemoveEnvFromApp call.

Comment (iamhopaul123) [reply internal/pkg/deploy/cloudformation/app.go:386]: > because we have to check in env_delete.go whether or not to empty the s3 buckets and ecr repos.

Maybe we can let `RemoveEnvFromApp` decide if we need to empty the s3 buckets and ecr repo before deleting the stackset instance?

File: internal/pkg/cli/job_delete.go
Comment (Lou1415926) [inline internal/pkg/cli/job_delete.go:306]: ty for doing the todo!

File: internal/pkg/deploy/cloudformation/app.go
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/app.go:380]: I feel like we should probably leave `Environments` out of the input params? Like I would expect  `RemoveEnvFromApp` to take only `envToRemove` and `App`.

File: internal/pkg/deploy/cloudformation/app.go
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/app.go:383]: Shall we update this comment accordingly?

File: internal/pkg/deploy/cloudformation/app.go
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/app.go]: nit: do we need to wrap this error?

File: internal/pkg/deploy/cloudformation/app.go
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/app.go:186]: Callout not for this PR but we probably need to refactor `deployAppConfig` to avoid using any boolean value (maybe split into two)

**Review by iamhopaul123 — APPROVED** [review]

File: internal/pkg/cli/pipeline_deploy_test.go
Comment (Lou1415926) [inline internal/pkg/cli/pipeline_deploy_test.go]: Do we have to use the mocks from the other packages 👀  Could we declare the interface as well as the mock in `cli` package?

File: internal/pkg/deploy/cloudformation/app.go
Comment (bvtujo) [reply internal/pkg/deploy/cloudformation/app.go:380]: Can you think of an easier way for the Cloudformation package to get access to the environments than passing them in here? I don't want to pass a `store` into the `CloudFormation` struct if I can avoid it; that feels like too much abstraction blending to me and it would be a first for this package.

Comment (bvtujo) [reply internal/pkg/deploy/cloudformation/app.go:380]: The CFN stacks we're dealing with don't have any metadata about where the environments are deployed. Typically we fetch information about the application with a config store or deploy store object, but again, passing a store to CFN feels like inverting the level of abstraction (and may lead to a circular dependency if we use a deploystore)

**Comment by bvtujo** [comment]
Applying a label to stop merging while we work through discussions

File: internal/pkg/deploy/cloudformation/app.go
Comment (Lou1415926) [inline internal/pkg/deploy/cloudformation/app.go]: nit:
```suggestion
	appConfig := stack.NewAppStackConfig(&deploy.CreateAppInput{
		Name:               opts.App.Name,
		AccountID:          opts.App.AccountID,
		DomainName:         opts.App.Domain,
		DomainHostedZoneID: opts.App.DomainHostedZoneID,
		Version:            opts.App.Version,
	})
```

File: internal/pkg/deploy/cloudformation/app.go
Comment (iamhopaul123) [reply internal/pkg/deploy/cloudformation/app.go:380]: Yeah totally agreed. I'm good for leaving it as is for now and probably refactor in the following PR if we can find a good way.

File: internal/pkg/deploy/cloudformation/app.go
Comment (Lou1415926) [inline internal/pkg/deploy/cloudformation/app.go:412]: I like that these are handled in the same function! It feels weird though that all these logics non-CFN are in the `deploy/cloudformation` package. Maybe this signals that we will need a refactor soon.

File: internal/pkg/deploy/cloudformation/app.go
Comment (Lou1415926) [inline internal/pkg/deploy/cloudformation/app.go:704]: Just making sure - if there is no instance in the said account + region, i take that we will still get an operation ID. What happens when we wait for that empty operation? Is that an instant return?

File: internal/pkg/deploy/cloudformation/app_test.go
Comment (Lou1415926) [inline internal/pkg/deploy/cloudformation/app_test.go:1409]: nit: we can add ⬇️  to be super clear that we want no such call. The other maintainers would be able to see the intention clearly.
```go
appStackSet.Expect().Describe("phonetool-infrastructure").Times(0)
appStackSet.Expect(). Update("phonetool-infrastructure").Times(0)

```

File: internal/pkg/deploy/cloudformation/app_test.go
Comment (Lou1415926) [inline internal/pkg/deploy/cloudformation/app_test.go:1491]: Samesies :ty

**Review by Lou1415926 — COMMENTED** [review]
lgtm! just a couple of nits & questions

File: internal/pkg/cli/pipeline_deploy_test.go
Comment (bvtujo) [reply internal/pkg/cli/pipeline_deploy_test.go]: I don't know what's going on with this package; for whatever reason my local mainline wouldn't build because of some interface issues. I'd prefer to remove this commit (and these pipeline changes) entirely. Probably will happen when I resolve merge conflicts.

File: internal/pkg/deploy/cloudformation/app.go
Comment (bvtujo) [reply internal/pkg/deploy/cloudformation/app.go:412]: Agreed, I think that perhaps this stuff needs to get pulled up one level into the deploy package. Or moved over into clideploy. It's probably worth putting all our deletion logic into that package as well as the deployment logic, now that I think about it.

File: internal/pkg/deploy/cloudformation/app.go
Comment (bvtujo) [reply internal/pkg/deploy/cloudformation/app.go:704]: Yeah, it instantly returns "succeeded". CFN views it as a successful delete even though nothing was deleted.

File: internal/pkg/cli/pipeline_deploy_test.go
Comment (bvtujo) [reply internal/pkg/cli/pipeline_deploy_test.go]: fixed by #5077; i'd like to get that merged to mainline before this PR

**Review by Lou1415926 — APPROVED** [review]
Applying a DNM so that this can be merged after the mock PR

File: internal/pkg/deploy/cloudformation/app.go
Comment (Lou1415926) [reply internal/pkg/deploy/cloudformation/app.go:186]: Quick question! Why do we need to call `deployAppConfig` which looks like is updating the stackset? 

My understanding so far is that stackset is only touched when we add/delete new regions & add/delete new workloads. Why do we need to update stackset when we remove a account?

Comment (Lou1415926) [reply internal/pkg/deploy/cloudformation/app.go:186]: oh I think I see now :homer-disappearing

We need to update the app stack so that we remove the account from delegation role; we need to update the stackset so that we remove the account from the metadata.

**Pull request merged**
